src/io/flutter/view/FlutterView.java

```diff
@@ -373,6 +383,109 @@ public void debugActive(@NotNull FlutterViewMessages.FlutterDebugEvent event) {
     }
   }
 
+  protected void handleJxBrowserInstalled(FlutterApp app, InspectorService inspectorService, ToolWindow toolWindow) {
+    final DevToolsManager devToolsManager = DevToolsManager.getInstance(app.getProject());
+
+    if (devToolsManager.hasInstalledDevTools()) {
+      addBrowserInspectorViewContent(app, inspectorService, toolWindow);
+    }
+    else {
+      presentLabel(toolWindow, INSTALLING_DEVTOOLS_LABEL);
+
+      final CompletableFuture<Boolean> result = devToolsManager.installDevTools();
+      result.thenAccept(succeeded -> {
+        if (succeeded) {
+          addBrowserInspectorViewContent(app, inspectorService, toolWindow);
+        } else {
+          // TODO(helin24): Handle with alternative instructions if devtools fails.
+          presentLabel(toolWindow, DEVTOOLS_FAILED_LABEL);
+        }
+      });
+    }
+  }
+
+  protected void handleJxBrowserInstallationInProgress(FlutterApp app, InspectorService inspectorService, ToolWindow toolWindow) {
+    final JxBrowserManager jxBrowserManager = JxBrowserManager.getInstance();
+    final DevToolsManager devToolsManager = DevToolsManager.getInstance(app.getProject());
+
+    presentLabel(toolWindow, INSTALLATION_IN_PROGRESS_LABEL);
+
+    // Start devtools while waiting for JxBrowser download.
+    devToolsManager.installDevTools();
+
+    if (jxBrowserManager.getStatus().equals(JxBrowserStatus.INSTALLED)) {
+      handleJxBrowserInstalled(app, inspectorService, toolWindow);
+    } else {
+      startJxBrowserInstallationWaitingThread(app, inspectorService, toolWindow);
+    }
+  }
+
+  protected void startJxBrowserInstallationWaitingThread(FlutterApp app, InspectorService inspectorService, ToolWindow toolWindow) {
+    ApplicationManager.getApplication().executeOnPooledThread(() -> {
+      waitForJxBrowserInstallation(app, inspectorService, toolWindow);
+    });
+  }
+
+  protected void waitForJxBrowserInstallation(FlutterApp app, InspectorService inspectorService, ToolWindow toolWindow) {
+    final JxBrowserManager jxBrowserManager = JxBrowserManager.getInstance();
+    // Periodically check for download to finish.
+    // TODO(helin24): Maybe set up JxBrowserManager as listener instead.
+    int waitSeconds = 1;
+    while (waitSeconds <= 128) {
```

Comment:
you should listen for an event from JxBrowserManager indicating that installation succeeded or failed rather than sleeping.

Reply:
you could also use CompletableFuture completing the future when JxBrowserManager is ready.